### PR TITLE
Zed IPA with HTTP Boot

### PR DESCRIPTION
**Pull request by fwiesel**
This build process relies more on the standard features of mkosi.

* `mkosi.build`: Instead of building and extracting the kernel and ramdisk with the ironic-python-agent integrated,
we only generates wheels of the submodules under `src/` and place them into `DESTDIR`. The use of submodules should ensure a fixed commit of each of the inputs. Additionally, the `requirements/upper-constraints.txt` are now honoured.
* `mkosi.postinst` now takes those wheels from the build step and installs them in the ipa virtual environment, which is then part of the root-disk
* `mkosi.fedora`: The changes cause mkosi to create a directory `image` with the filesystem containing the IPA, the ramdisk `image.initrd`, and the kernel `image.vmlinuz`. The kernel is the standard fedora one.
* `mkosi.extra`: Contains all files to be copied into both the build and the final image, that includes the `mkosi.extra/etc/dracut.conf.d/99-ipa.conf` which contains the configuration for the ramdisk, and adds the `livenet` module, which allows us to load the root-disk over http by adding the kernel command line `root=live:https://repo.example.com/ipa.squashfs`

To improve the turnaround time in testing `make test` will spin up a VM with the unsecured api exposed at port 9999 on the host of the VM.

### CODE REVIEW

File: mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-getty.conf [hunk 1]
```diff
@@ -1,3 +0,0 @@
-[Service]
-ExecStart=
-ExecStart=-/sbin/agetty --skip-login --login-options root --noclear %I $TERM
```
Comment (killermoehre) [inline mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-getty.conf:3]: Why was the auto-login removed?

File: mkosi.extra/etc/systemd/journal.conf.d/augment-journald.conf [hunk 2]
```diff
@@ -7,5 +7,5 @@ RateLimitIntervalSec=0
 RateLimitBurst=0
 RuntimeKeepFree=5%
 ForwardToConsole=true
-TTYPath=/dev/tty1
+TTYPath=/dev/ttyS0
```
Comment (killermoehre) [inline mkosi.extra/etc/systemd/journal.conf.d/augment-journald.conf:10]: Is this only viable for testing? What about actual systems in hardware?

File: mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-serial-getty.conf [hunk 3]
```diff
@@ -1,3 +0,0 @@
-[Service]
-ExecStart=
-ExecStart=-/sbin/agetty --skip-login --login-options root --keep-baud 115200,38400,9600 %I $TERM
```
Comment (killermoehre) [inline mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-serial-getty.conf:3]: Why was this auto-login removed?

File: mkosi.postinst [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+[ "$1" == "final" ] || exit 0
+
+systemctl enable systemd-networkd
+systemctl enable ironic-python-agent
+systemctl set-default -f multi-user.target
+systemctl disable -f graphical.target
```
Comment (killermoehre) [inline mkosi.postinst:8]: Please move this steps directly into `mkosi.extra`, as all of those are just symlinks in the file system.

File: mkosi.finalize [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+[ "$1" == "final" ] || exit 0
+
+exit 0
+
+cd $OUTPUTDIR
+
+mount \
+ -o loop \
+ image.root.raw \
+ image.root.dir
```
Comment (killermoehre) [inline mkosi.finalize:13]: What happens if `image.root.dir` does not exist? If this directory created than?

File: mkosi.finalize [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+set -Eeuo pipefail
+
+[ "$1" == "final" ] || exit 0
+
+exit 0
```
Comment (killermoehre) [inline mkosi.finalize:6]: Everything after this `exit 0` is ignored. Please either remove the line or the file altogether.

**Review by killermoehre — CHANGES_REQUESTED** [review]

File: mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-getty.conf (hunk 1 above)
Comment (fwiesel) [reply mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-getty.conf:3]: That is now solved by the [config-option](https://github.com/sapcc/fedIPA/pull/21/files#diff-eecc9f3923757c5c1e627ce59942b0160198e2daa7ff4ab810d6d3cd5fae00a4R22).

File: mkosi.extra/etc/systemd/journal.conf.d/augment-journald.conf (hunk 2 above)
Comment (fwiesel) [reply mkosi.extra/etc/systemd/journal.conf.d/augment-journald.conf:10]: To my knowledge, we have configured all systems to provide SOL, so the log should be available there too.
But I can also revert it back.

File: mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-serial-getty.conf (hunk 3 above)
Comment (fwiesel) [reply mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-serial-getty.conf:3]: See: https://github.com/sapcc/fedIPA/pull/21#discussion_r1002957012

File: mkosi.postinst (hunk 4 above)
Comment (fwiesel) [reply mkosi.postinst:8]: The problem is they are symlinks installed according to the internal logic of systemd and depending on the packages installed in the prior step.
More importantly, I cannot remove links with `mkosi.extra`, only add.

File: mkosi.finalize (hunk 5 above)
Comment (fwiesel) [reply mkosi.finalize]: No, it is some left-over before I moved the squashfs file creation. completely to the Makefile.
I'll remove the file.

File: mkosi.finalize (hunk 6 above)
Comment (fwiesel) [reply mkosi.finalize]: I've removed the file

File: mkosi.extra/etc/systemd/journal.conf.d/augment-journald.conf (hunk 2 above)
Comment (fwiesel) [reply mkosi.extra/etc/systemd/journal.conf.d/augment-journald.conf]: I've changed it back.

File: mkosi.postinst (hunk 4 above)
Comment (killermoehre) [reply mkosi.postinst:8]: You can `mask` services to not start by symlinking to `/dev/null`.

File: mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-getty.conf (hunk 1 above)
Comment (killermoehre) [reply mkosi.extra/usr/lib/dracut/modules.d/01setup-ironic-python-agent/20-autologin-getty.conf:3]: I see the config option. Very well. Simple failing is not good, of course.

File: mkosi.postinst (hunk 4 above)
Comment (killermoehre) [reply mkosi.postinst:8]: Also, if `default.target` points as symlink to `multi-user.target`, then the `graphical.target` won't be started (except if explicitly requested).

Comment (fwiesel) [reply mkosi.postinst:8]: Which shows your knowledge of systemd exceeds mine, and explains why I think `systemctl disable -f graphical.target`  is much more readable and easy to understand, than having a symlink in `mkosi.extra/etc/systemd/...` pointing to `/dev/null`.

In the end, the difference is stylistic, isn't it? I think, neither solution is better in any objective metric.

Comment (fwiesel) [reply mkosi.postinst]: I've removed `systemctl disable -f graphical.target`.

**Review by killermoehre — APPROVED** [review]

**Review by sandzwerg — APPROVED** [review]
fine with me if silvio is fine with it

**Pull request merged**